egs/librispeech/ASR/transducer_stateless/alignment.py

```diff
@@ -0,0 +1,264 @@
+# Copyright    2022  Xiaomi Corp.        (authors: Fangjun Kuang)
+#
+# See ../../../../LICENSE for clarification regarding multiple authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from dataclasses import dataclass
+from typing import Iterator, List, Optional
+
+import sentencepiece as spm
+import torch
+from model import Transducer
+
+# The force alignment problem can be formulated as finding
+# a path in a rectangular lattice, where the path starts
+# from the lower left corner and ends at the upper right
+# corner. The horizontal axis of the lattice is `t` (representing
+# acoustic frame indexes) and the vertical axis is `u` (representing
+# BPE tokens of the transcript).
+#
+# Beam search is used to find the path with the
+# highest log probabilities.
+#
+# It assumes the maximum number of symbols that can be
+# emitted per frame is 1. You can use `--modified-transducer-prob`
+# from `./train.py` to train a model that satisfies this assumption.
+
+
+# AlignItem is a node in the lattice, where its
+# len(ys) equals to `t` and pos_u is the u coordinate
+# in the lattice.
+@dataclass
+class AlignItem:
+    # log prob of this item originating from the start item
```

Comment:
when you say "start item", do you mean the preceding item (i.e. is it the log-prob p(this item | preceding item), or is it some kind of total log-prob from the start node of the lattice to here?

Reply:
I mean the total log prob from the starting node.

`AlignItem` can be considered as the ending point of a path originated from the starting node and the `log_probs` is the tot_log_prob of this path.